File: algobet/api/routers/schedules.py
"""API router for scheduled task management."""

from datetime import datetime
from typing import Any

from fastapi import APIRouter, BackgroundTasks, Depends, HTTPException, Query, status
from pydantic import BaseModel, Field
from sqlalchemy.orm import Session

from algobet.api.dependencies import get_db
from algobet.models import ScheduledTask, TaskExecution
from algobet.services.scheduler_service import SchedulerService

router = APIRouter(tags=["schedules"])


# Request/Response Schemas
class ScheduleCreateRequest(BaseModel):
    """Request schema for creating a scheduled task."""

    name: str = Field(
        ...,
        description="Unique name for the scheduled task",
        min_length=1,
        max_length=255,
    )
    task_type: str = Field(..., description="Type of task to execute")
    cron_expression: str = Field(
        ..., description="Cron expression for scheduling (e.g., '0 6 * * *')"
    )
    parameters: dict[str, Any] = Field(
        default_factory=dict, description="Task-specific parameters"
    )
    description: str | None = Field(None, description="Human-readable description")
    is_active: bool = Field(True, description="Whether the schedule is active")


class ScheduleUpdateRequest(BaseModel):
    """Request schema for updating a scheduled task."""

    cron_expression: str | None = Field(None, description="Updated cron expression")
    parameters: dict[str, Any] | None = Field(
        None, description="Updated task parameters"
    )
    description: str | None = Field(None, description="Updated description")
    is_active: bool | None = Field(None, description="Updated active status")


class ScheduleResponse(BaseModel):
    """Response schema for a scheduled task."""

    id: int = Field(..., description="Unique identifier")
    name: str = Field(..., description="Task name")
    task_type: str = Field(..., description="Task type")
    cron_expression: str = Field(..., description="Cron expression")
    is_active: bool = Field(..., description="Active status")
    parameters: dict[str, Any] = Field(..., description="Task parameters")
    description: str | None = Field(None, description="Task description")
    created_at: datetime = Field(..., description="Creation timestamp")
    updated_at: datetime = Field(..., description="Last update timestamp")

    class Config:
        from_attributes = True


class TaskExecutionResponse(BaseModel):
    """Response schema for a task execution."""

    id: int = Field(..., description="Execution ID")
    task_id: int = Field(..., description="Parent task ID")
    status: str = Field(..., description="Execution status")
    started_at: datetime = Field(..., description="Start timestamp")
    completed_at: datetime | None = Field(None, description="Completion timestamp")
    result: dict[str, Any] | None = Field(None, description="Execution result")
    error_message: str | None = Field(None, description="Error message if failed")

    class Config:
        from_attributes = True


class ScheduleListResponse(BaseModel):
    """Response schema for listing schedules."""

    schedules: list[ScheduleResponse] = Field(..., description="List of schedules")
    total: int = Field(..., description="Total count")


class ExecutionHistoryResponse(BaseModel):
    """Response schema for execution history."""

    executions: list[TaskExecutionResponse] = Field(
        ..., description="List of executions"
    )
    total: int = Field(..., description="Total count")


class ScheduleTaskTypesResponse(BaseModel):
    """Response schema for available task types."""

    task_types: list[str] = Field(..., description="Available task types")


@router.post("/", response_model=ScheduleResponse, status_code=status.HTTP_201_CREATED)
async def create_schedule(
    request: ScheduleCreateRequest,
    db: Session = Depends(get_db),
) -> ScheduledTask:
    """Create a new scheduled task.

    Args:
        request: Schedule creation request
        db: Database session

    Returns:
        Created scheduled task

    Raises:
        HTTPException: If task creation fails
    """
    try:
        scheduler = SchedulerService(db)
        task = scheduler.create_schedule(
            name=request.name,
            task_type=request.task_type,
            cron_expression=request.cron_expression,
            parameters=request.parameters,
            description=request.description,
            is_active=request.is_active,
        )
        return task
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail=str(e),
        ) from e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to create schedule: {str(e)}",
        ) from e


@router.get("/", response_model=ScheduleListResponse)
async def list_schedules(
    task_type: str | None = None,
    is_active: bool | None = None,
    db: Session = Depends(get_db),
) -> ScheduleListResponse:
    """List all scheduled tasks with optional filtering.

    Args:
        task_type: Optional filter by task type
        is_active: Optional filter by active status
        db: Database session

    Returns:
        List of scheduled tasks
    """
    scheduler = SchedulerService(db)
    schedules = scheduler.list_schedules(task_type=task_type, is_active=is_active)

    # Convert ORM models to Pydantic models
    schedule_responses = [ScheduleResponse.model_validate(s) for s in schedules]

    return ScheduleListResponse(
        schedules=schedule_responses,
        total=len(schedules),
    )


@router.get("/task-types", response_model=ScheduleTaskTypesResponse)
async def get_task_types(
    db: Session = Depends(get_db),
) -> ScheduleTaskTypesResponse:
    """Get all available task types.

    Args:
        db: Database session

    Returns:
        List of available task types
    """
    task_types = SchedulerService.get_task_types()
    return ScheduleTaskTypesResponse(task_types=task_types)


@router.get("/{schedule_id}", response_model=ScheduleResponse)
async def get_schedule(
    schedule_id: int,
    db: Session = Depends(get_db),
) -> ScheduledTask:
    """Get a scheduled task by ID.

    Args:
        schedule_id: Schedule ID
        db: Database session

    Returns:
        Scheduled task details

    Raises:
        HTTPException: If schedule not found
    """
    scheduler = SchedulerService(db)
    task = scheduler.get_schedule(schedule_id)

    if not task:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=f"Schedule with ID {schedule_id} not found",
        )

    return task


@router.patch("/{schedule_id}", response_model=ScheduleResponse)
async def update_schedule(
    schedule_id: int,
    request: ScheduleUpdateRequest,
    db: Session = Depends(get_db),
) -> ScheduledTask:
    """Update a scheduled task.

    Args:
        schedule_id: Schedule ID
        request: Update request
        db: Database session

    Returns:
        Updated scheduled task

    Raises:
        HTTPException: If schedule not found or update fails
    """
    try:
        scheduler = SchedulerService(db)
        task = scheduler.update_schedule(
            schedule_id=schedule_id,
            cron_expression=request.cron_expression,
            parameters=request.parameters,
            description=request.description,
            is_active=request.is_active,
        )
        return task
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to update schedule: {str(e)}",
        ) from e


@router.delete("/{schedule_id}", status_code=status.HTTP_204_NO_CONTENT)
async def delete_schedule(
    schedule_id: int,
    db: Session = Depends(get_db),
) -> None:
    """Delete a scheduled task.

    Args:
        schedule_id: Schedule ID
        db: Database session

    Raises:
        HTTPException: If schedule not found
    """
    try:
        scheduler = SchedulerService(db)
        scheduler.delete_schedule(schedule_id)
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e


@router.post("/{schedule_id}/run", response_model=TaskExecutionResponse)
async def run_schedule_now(
    schedule_id: int,
    background_tasks: BackgroundTasks,
    db: Session = Depends(get_db),
) -> TaskExecution:
    """Execute a scheduled task immediately.

    Args:
        schedule_id: Schedule ID
        background_tasks: FastAPI background tasks
        db: Database session

    Returns:
        Task execution record

    Raises:
        HTTPException: If schedule not found or execution fails
    """
    try:
        scheduler = SchedulerService(db)
        execution = scheduler.execute_task(schedule_id)
        return execution
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e
    except Exception as e:
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Failed to execute schedule: {str(e)}",
        ) from e


@router.get("/{schedule_id}/history", response_model=ExecutionHistoryResponse)
async def get_execution_history(
    schedule_id: int,
    limit: int = Query(
        100, ge=1, le=1000, description="Maximum number of executions to return"
    ),
    db: Session = Depends(get_db),
) -> ExecutionHistoryResponse:
    """Get execution history for a scheduled task.





    Args:


        schedule_id: Schedule ID


        limit: Maximum number of executions to return


        db: Database session





    Returns:


        Execution history





    Raises:


        HTTPException: If schedule not found


    """

    try:
        scheduler = SchedulerService(db)

        executions = scheduler.get_execution_history(schedule_id, limit=limit)

        # Convert ORM models to Pydantic models

        execution_responses = [
            TaskExecutionResponse.model_validate(e) for e in executions
        ]

        return ExecutionHistoryResponse(
            executions=execution_responses,
            total=len(executions),
        )

    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e


@router.get("/{schedule_id}/last-execution", response_model=TaskExecutionResponse)
async def get_last_execution(
    schedule_id: int,
    db: Session = Depends(get_db),
) -> TaskExecution:
    """Get the most recent execution for a scheduled task.

    Args:
        schedule_id: Schedule ID
        db: Database session

    Returns:
        Last execution record

    Raises:
        HTTPException: If schedule not found or no executions exist
    """
    try:
        scheduler = SchedulerService(db)
        execution = scheduler.get_last_execution(schedule_id)

        if not execution:
            raise HTTPException(
                status_code=status.HTTP_404_NOT_FOUND,
                detail=f"No execution history found for schedule {schedule_id}",
            )

        return execution
    except ValueError as e:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND,
            detail=str(e),
        ) from e

File: frontend/lib/api/schedules.ts
/**
 * API client functions for schedule management
 */

import { apiGet, apiPost, apiPatch, apiDelete, buildQueryString } from './client'
import { z } from 'zod'

// Zod schemas for runtime validation
export const ScheduledTaskSchema = z.object({
  id: z.number(),
  name: z.string(),
  task_type: z.string(),
  cron_expression: z.string(),
  is_active: z.boolean(),
  parameters: z.record(z.unknown()),
  description: z.string().nullable(),
  created_at: z.string(),
  updated_at: z.string(),
})

export const TaskExecutionSchema = z.object({
  id: z.number(),
  task_id: z.number(),
  status: z.enum(['pending', 'running', 'completed', 'failed']),
  started_at: z.string(),
  completed_at: z.string().nullable(),
  duration: z.number().nullable(),
  result: z.record(z.unknown()).nullable(),
  error_message: z.string().nullable(),
})

export const ScheduleListResponseSchema = z.object({
  schedules: z.array(ScheduledTaskSchema),
  total: z.number(),
})

export const ExecutionHistoryResponseSchema = z.object({
  executions: z.array(TaskExecutionSchema),
  total: z.number(),
})

export const ScheduleCreateRequestSchema = z.object({
  name: z.string(),
  task_type: z.string(),
  cron_expression: z.string(),
  parameters: z.record(z.unknown()).optional(),
  description: z.string().optional(),
  is_active: z.boolean().optional(),
})

export const ScheduleUpdateRequestSchema = z.object({
  cron_expression: z.string().optional(),
  parameters: z.record(z.unknown()).optional(),
  description: z.string().optional(),
  is_active: z.boolean().optional(),
})

// Types derived from schemas
export type ScheduledTask = z.infer<typeof ScheduledTaskSchema>
export type TaskExecution = z.infer<typeof TaskExecutionSchema>
export type ScheduleListResponse = z.infer<typeof ScheduleListResponseSchema>
export type ExecutionHistoryResponse = z.infer<typeof ExecutionHistoryResponseSchema>
export type ScheduleCreateRequest = z.infer<typeof ScheduleCreateRequestSchema>
export type ScheduleUpdateRequest = z.infer<typeof ScheduleUpdateRequestSchema>

/**
 * Get all scheduled tasks
 */
export async function getSchedules(options?: {
  task_type?: string
  is_active?: boolean
}): Promise<ScheduleListResponse> {
  const params: Record<string, unknown> = {}
  if (options?.task_type) params.task_type = options.task_type
  if (options?.is_active !== undefined) params.is_active = String(options.is_active)
  
  const queryString = buildQueryString(params)
  return apiGet(`/schedules${queryString}`, ScheduleListResponseSchema)
}

/**
 * Get a specific scheduled task by ID
 */
export async function getSchedule(scheduleId: number): Promise<ScheduledTask> {
  return apiGet(`/schedules/${scheduleId}`, ScheduledTaskSchema)
}

/**
 * Get available task types
 */
export async function getTaskTypes(): Promise<string[]> {
  const schema = z.object({ task_types: z.array(z.string()) })
  const result = await apiGet('/schedules/task-types', schema)
  return result.task_types
}

/**
 * Create a new scheduled task
 */
export async function createSchedule(request: ScheduleCreateRequest): Promise<ScheduledTask> {
  return apiPost('/schedules', request, ScheduledTaskSchema)
}

/**
 * Update a scheduled task
 */
export async function updateSchedule(
  scheduleId: number,
  request: ScheduleUpdateRequest
): Promise<ScheduledTask> {
  return apiPatch(`/schedules/${scheduleId}`, request, ScheduledTaskSchema)
}

/**
 * Delete a scheduled task
 */
export async function deleteSchedule(scheduleId: number): Promise<void> {
  return apiDelete(`/schedules/${scheduleId}`)
}

/**
 * Run a scheduled task immediately
 */
export async function runScheduleNow(scheduleId: number): Promise<TaskExecution> {
  return apiPost(`/schedules/${scheduleId}/run`, {}, TaskExecutionSchema)
}

/**
 * Get execution history for a scheduled task
 */
export async function getExecutionHistory(
  scheduleId: number,
  limit: number = 100
): Promise<ExecutionHistoryResponse> {
  return apiGet(`/schedules/${scheduleId}/history?limit=${limit}`, ExecutionHistoryResponseSchema)
}

/**
 * Get the last execution for a scheduled task
 */
export async function getLastExecution(scheduleId: number): Promise<TaskExecution> {
  return apiGet(`/schedules/${scheduleId}/last-execution`, TaskExecutionSchema)
}